git/index/typ.py

```diff
@@ -48,12 +51,14 @@ def __init__(self, paths: Sequence[PathLike]) -> None:
         """
         self.paths = paths
 
-    def __call__(self, stage_blob: Blob) -> bool:
-        path = stage_blob[1].path
-        for p in self.paths:
-            if path.startswith(p):
+    def __call__(self, stage_blob: Tuple[StageType, Blob]) -> bool:
+        blob_pathlike: PathLike = stage_blob[1].path
+        blob_path: Path = blob_pathlike if isinstance(blob_pathlike, Path) else Path(blob_pathlike)
+        for pathlike in self.paths:
+            path: Path = pathlike if isinstance(pathlike, Path) else Path(pathlike)
+            # TODO: Change to use `PosixPath.is_relative_to` once Python 3.8 is no longer supported.
+            if all(i == j for i, j in zip(path.parts, blob_path.parts)):
```

Comment:
Is this not another way of writing `path1 == path2`, since all path components (here `.parts`)  have to be equivalent? The benefit of doing so of course is that it ignores which path separator is used, but I am having trouble to understand how this checks that `blob_path` starts with one of `self.paths`.

Reply:
I think this should work because `zip` stops when the shortest of the arguments is exhausted. So it is only checking if _all_ the path components are equivalent in the case that the length of the `parts` are equal. But, when one path has more parts than the other, then it is checking that the longer path starts with the shorter path.

Example:
```python3
def startswith(a, b): return all(i == j for i, j in zip(a, b))

a = [1, 2, 3]
b = [1, 2]
assert startswith(a, b)
```

Reply:
Thanks for the explanation, I see how this works now and why this is correct. The alternative has to remain a TODO until it becomes available, and looking at the code I am happy to suggest to not ever go there - the try:except clause seems like a deal-breaker to me. It's entirely up to you though.

Reply:
Actually, let me undo what I last said. I once again think this implementation is incorrect, as these paths are not equivalent.

```python
base = [1, 2, 3]
added_path = [1, 2]
```

The above would match even though `added_path` is not contained in base.

Reply:
Good catch, I pushed a change that addresses this and adds a test too.